packages/websocket-client/src/index.ts

```diff
@@ -0,0 +1,39 @@
+import * as WebSocket from 'isomorphic-ws';
+
+import { spawn, ensure, Resource } from '@effection/core'
+import { createQueue, Subscription } from '@effection/subscription';
+import { on, once } from '@effection/events';
+
+interface WebsocketClient<TIncoming = unknown, TOutgoing = TIncoming> extends Subscription<TIncoming> {
+  send(message: TOutgoing): void;
+}
+
+export function createWebSocketClient<TIncoming = unknown, TOutgoing = TIncoming>(url: string): Resource<WebsocketClient<TIncoming, TOutgoing>> {
+  return {
+    *init() {
+      let socket = new WebSocket(url);
+
+      yield once(socket, 'open');
+
+      let queue = createQueue<TIncoming>();
+
+      yield spawn(on<{ data: string }>(socket, 'message').forEach((message) => { queue.send(JSON.parse(message.data)) }));
+      yield spawn(function*() {
+        let { wasClean, code, reason } = yield once(socket, 'close');
+        if(wasClean) {
+          queue.close();
+        } else {
+          throw new Error(`websocket server closed connection unexpectedly: [${code}] ${reason}`);
+        }
+      });
+      yield ensure(() => { socket.close() });
+
+      return {
+        ...queue.subscription,
+        send(message: TOutgoing) {
+          socket.send(JSON.stringify(message));
+        }
+      }
+    }
+  }
+}
```

Comment:
If we create a single top level "resource task", then it will work well with our visual debugger in the future, right?

```ts
export function createWebSocketClient<TIncoming = unknown, TOutgoing = TIncoming>(url: string): Resource<WebsocketClient<TIncoming, TOutgoing>> {
   return {
     *init() {
       let socket = new WebSocket(url);
  
       yield spawn(function*() {
         // not yet, but soon!
         // yield setLabels({ name: `websocket client`, url })
         try {
           let queue = createQueue<TIncoming>();
           yield spawn(function*() {
             let { wasClean, code, reason } = yield once(socket, 'close');
             if(wasClean) {
               queue.close();
             } else {
               throw new Error(`websocket server closed connection unexpectedly: [${code}] ${reason}`);
             }
           });
  
           yield on<{ data: string }>(socket, 'message').forEach((message) => { queue.send(JSON.parse(message.data)) });
         } finally {
           socket.close();
         }
       });

       yield once(socket, 'open');

       return {
         ...queue.subscription,
         send(message: TOutgoing) {
           socket.send(JSON.stringify(message));
         }
       }
     }
   }
 }
```